File: rl_mlx/features.py
"""与 web/src/bot/features.js 一致的数值特征（常数来自 shared/game_rules.json）。"""

from __future__ import annotations

import math
from typing import Sequence

import numpy as np

from .game_rules import FEATURE_ENCODING, rl_bonus_block_icons
from .shapes_data import get_all_shapes

_ENC = FEATURE_ENCODING
_AF = float(_ENC.get("almostFullLineRatio", 0.78))
_DOCK = float(_ENC.get("dockSlots", 3))
_AN = dict(_ENC.get("actionNorm") or {})
_MAX_GRID = int(_ENC.get("maxGridWidth", 8))
_DOCK_MASK_SIDE = int(_ENC.get("dockMaskSide", 5))
_DOCK_SLOTS = int(_ENC.get("dockSlots", 3))
_SCALAR_DIM = int(_ENC.get("stateScalarDim", 15))
_N_COLORS = int(_ENC.get("colorCount", 8))
_RL_BONUS_ICONS = rl_bonus_block_icons()

STATE_FEATURE_DIM = int(_ENC["stateDim"])
ACTION_FEATURE_DIM = int(_ENC["actionDim"])
PHI_DIM = int(_ENC.get("phiDim") or STATE_FEATURE_DIM + ACTION_FEATURE_DIM)

_GRID_FLAT = _MAX_GRID * _MAX_GRID
_DOCK_FLAT = _DOCK_SLOTS * _DOCK_MASK_SIDE * _DOCK_MASK_SIDE
_expected_state = _SCALAR_DIM + _GRID_FLAT + _DOCK_FLAT
if STATE_FEATURE_DIM != _expected_state:
    raise ValueError(
        f"game_rules featureEncoding.stateDim={STATE_FEATURE_DIM} 与 "
        f"stateScalarDim+grid²+dock 期望 {_expected_state} 不一致"
    )


def _std_dev(arr: Sequence[float]) -> float:
    if not arr:
        return 0.0
    m = sum(arr) / len(arr)
    v = sum((x - m) ** 2 for x in arr) / len(arr)
    return math.sqrt(v)


def _can_place_shape_at(grid, shape: list[list[int]], gx: int, gy: int) -> bool:
    for y, row in enumerate(shape):
        for x, v in enumerate(row):
            if not v:
                continue
            px, py = gx + x, gy + y
            if px < 0 or px >= grid.size or py < 0 or py >= grid.size:
                return False
            if grid.cells[py][px] is not None:
                return False
    return True


def count_unfillable_cells(grid, shapes: list[dict] | None = None) -> int:
    n = grid.size
    coverable = [[False] * n for _ in range(n)]
    for shape_info in shapes or get_all_shapes():
        shape = shape_info.get("data") if isinstance(shape_info, dict) else shape_info
        if not shape:
            continue
        h = len(shape)
        w = len(shape[0]) if h else 0
        for gy in range(0, n - h + 1):
            for gx in range(0, n - w + 1):
                if not _can_place_shape_at(grid, shape, gx, gy):
                    continue
                for sy, row in enumerate(shape):
                    for sx, v in enumerate(row):
                        if v:
                            coverable[gy + sy][gx + sx] = True
    return sum(
        1
        for y in range(n)
        for x in range(n)
        if grid.cells[y][x] is None and not coverable[y][x]
    )


def _encode_grid_occupancy(grid) -> np.ndarray:
    """行主序 flatten，棋盘置于左上角，不足 max 则右侧/下侧填 0。"""
    n = grid.size
    out = np.zeros((_MAX_GRID, _MAX_GRID), dtype=np.float32)
    for y in range(min(n, _MAX_GRID)):
        row = grid.cells[y]
        for x in range(min(n, _MAX_GRID)):
            out[y, x] = 1.0 if row[x] is not None else 0.0
    return out.reshape(-1)


def _encode_shape_mask(shape: list[list[int]]) -> np.ndarray:
    """将多连块置于 dockMaskSide×dockMaskSide 画布中心，行主序 1/0。"""
    h = len(shape)
    w = len(shape[0]) if h else 0
    side = _DOCK_MASK_SIDE
    canvas = np.zeros((side, side), dtype=np.float32)
    if h == 0 or w == 0:
        return canvas.reshape(-1)
    off_y = max(0, (side - h) // 2)
    off_x = max(0, (side - w) // 2)
    for py in range(h):
        for px in range(w):
            if shape[py][px]:
                cy, cx = off_y + py, off_x + px
                if 0 <= cy < side and 0 <= cx < side:
                    canvas[cy, cx] = 1.0
    return canvas.reshape(-1)


def _encode_dock_spatial(dock: list[dict]) -> np.ndarray:
    parts: list[np.ndarray] = []
    for i in range(_DOCK_SLOTS):
        if i < len(dock) and not dock[i].get("placed"):
            parts.append(_encode_shape_mask(dock[i]["shape"]))
        else:
            parts.append(np.zeros(_DOCK_MASK_SIDE * _DOCK_MASK_SIDE, dtype=np.float32))
    return np.concatenate(parts, axis=0)


def _encode_color_summary(grid, dock: list[dict]) -> np.ndarray:
    n = grid.size
    area = max(n * n, 1)
    color_counts = np.zeros(_N_COLORS, dtype=np.float32)
    mono_line_potential = np.zeros(_N_COLORS, dtype=np.float32)
    for y in range(n):
        for x in range(n):
            c = grid.cells[y][x]
            if c is not None and 0 <= int(c) < _N_COLORS:
                color_counts[int(c)] += 1.0
    for c in range(_N_COLORS):
        best = 0
        for y in range(n):
            row = grid.cells[y]
            if all(v is None or v == c for v in row):
                best = max(best, sum(1 for v in row if v == c))
        for x in range(n):
            col = [grid.cells[y][x] for y in range(n)]
            if all(v is None or v == c for v in col):
                best = max(best, sum(1 for v in col if v == c))
        mono_line_potential[c] = best / max(n, 1)
    dock_colors = np.zeros(_DOCK_SLOTS, dtype=np.float32)
    denom = max(_N_COLORS - 1, 1)
    for i in range(_DOCK_SLOTS):
        if i < len(dock) and not dock[i].get("placed"):
            dock_colors[i] = float(dock[i].get("color_idx", dock[i].get("colorIdx", 0))) / denom
    return np.concatenate([color_counts / area, mono_line_potential, dock_colors], axis=0)


def extract_state_features(grid, dock: list[dict]) -> np.ndarray:
    n = grid.size
    area = n * n
    filled = 0
    row_fill: list[float] = []
    for y in range(n):
        r = sum(1 for x in range(n) if grid.cells[y][x] is not None)
        filled += int(r)
        row_fill.append(r / n)

    col_fill: list[float] = []
    for x in range(n):
        c = sum(1 for y in range(n) if grid.cells[y][x] is not None)
        col_fill.append(c / n)

    max_row = max(row_fill) if row_fill else 0.0
    min_row = min(row_fill) if row_fill else 0.0
    max_col = max(col_fill) if col_fill else 0.0
    min_col = min(col_fill) if col_fill else 0.0

    almost_full_rows = sum(1 for rf in row_fill if _AF <= rf < 1)
    almost_full_cols = sum(1 for cf in col_fill if _AF <= cf < 1)
    unplaced = sum(1 for b in dock if not b["placed"]) / _DOCK

    base_scalars = [
            filled / area,
            max_row,
            min_row,
            max_col,
            min_col,
            almost_full_rows / n,
            almost_full_cols / n,
            unplaced,
            sum(row_fill) / n,
            sum(col_fill) / n,
            _std_dev(row_fill),
            _std_dev(col_fill),
            max_row - min_row,
            max_col - min_col,
            (almost_full_rows + almost_full_cols) / (2 * n),
            min(count_unfillable_cells(grid) / float(_AN.get("maxHoles", 16)), 1.0),
            0.0,  # row transitions placeholder
            0.0,  # col transitions placeholder
            0.0,  # wells placeholder
            0.0,  # close1 placeholder
            0.0,  # close2 placeholder
            0.0,  # mobility placeholder
            filled / area,
        ]
    scalars = np.array(base_scalars + _encode_color_summary(grid, dock).tolist(), dtype=np.float32)
    if scalars.shape[0] != _SCALAR_DIM:
        raise ValueError(f"标量段长度 {scalars.shape[0]} != stateScalarDim {_SCALAR_DIM}")

    grid_flat = _encode_grid_occupancy(grid)
    dock_flat = _encode_dock_spatial(dock)
    return np.concatenate([scalars, grid_flat, dock_flat], axis=0)


def extract_action_features(
    state_feat: np.ndarray,
    block_idx: int,
    gx: int,
    gy: int,
    shape: list[list[int]],
    would_clear: int,
    grid_size: int,
    grid=None,
    dock: list[dict] | None = None,
    color_idx: int = 0,
) -> np.ndarray:
    cells = sum(1 for row in shape for c in row if c)
    h = len(shape)
    w = len(shape[0])
    div_b = float(_AN.get("maxBlockIndex", 3))
    div_sh = float(_AN.get("shapeSpan", 5))
    div_cells = float(_AN.get("maxCells", 10))
    div_clr = float(_AN.get("maxClearsHint", 5))
    unplaced_after = sum(1 for b in (dock or []) if not b.get("placed")) - 1
    multi_clear, bonus_line, perfect_clear = _clear_payoff_features(grid, shape, gx, gy, color_idx, div_clr)
    action_part = np.array(
        [
            block_idx / div_b,
            gx / grid_size,
            gy / grid_size,
            w / div_sh,
            h / div_sh,
            cells / div_cells,
            would_clear / div_clr,
            0.0,
            max(0, unplaced_after) / 3.0,
            0.0,
            max(gy + h, 0) / grid_size,
            0.0,
            multi_clear,
            bonus_line,
            perfect_clear,
        ],
        dtype=np.float32,
    )
    if action_part.shape[0] != ACTION_FEATURE_DIM:
        raise ValueError(f"动作特征长度 {action_part.shape[0]} != actionDim {ACTION_FEATURE_DIM}")
    return np.concatenate([state_feat, action_part], axis=0)


def _line_is_bonus(vals: list[int | None]) -> bool:
    if not vals or vals[0] is None:
        return False
    first = int(vals[0])
    if _RL_BONUS_ICONS:
        icon0 = _RL_BONUS_ICONS[first % len(_RL_BONUS_ICONS)]
        return all(v is not None and _RL_BONUS_ICONS[int(v) % len(_RL_BONUS_ICONS)] == icon0 for v in vals)
    return all(v == first for v in vals)


def _clear_payoff_features(
    grid,
    shape: list[list[int]],
    gx: int,
    gy: int,
    color_idx: int,
    div_clr: float,
) -> tuple[float, float, float]:
    if grid is None:
        return 0.0, 0.0, 0.0
    cells = [row[:] for row in grid.cells]
    for sy, row in enumerate(shape):
        for sx, v in enumerate(row):
            if v:
                cells[gy + sy][gx + sx] = int(color_idx)
    n = grid.size
    full_rows = [y for y in range(n) if all(c is not None for c in cells[y])]
    full_cols = [x for x in range(n) if all(cells[y][x] is not None for y in range(n))]
    clears = len(full_rows) + len(full_cols)
    if clears <= 0:
        return 0.0, 0.0, 0.0
    bonus = 0
    for y in full_rows:
        if _line_is_bonus(cells[y]):
            bonus += 1
    for x in full_cols:
        if _line_is_bonus([cells[y][x] for y in range(n)]):
            bonus += 1
    for y in full_rows:
        for x in range(n):
            cells[y][x] = None
    for x in full_cols:
        for y in range(n):
            cells[y][x] = None
    perfect = 1.0 if all(c is None for row in cells for c in row) else 0.0
    multi = min(max(clears - 1, 0) / max(div_clr - 1.0, 1.0), 1.0)
    return float(multi), float(min(bonus / max(div_clr, 1.0), 1.0)), perfect


def build_phi_batch(sim, legal: list[dict]) -> tuple[np.ndarray, np.ndarray]:
    """would_clear 与 web 端 countClearsIfPlaced 一致。sim: OpenBlockSimulator。"""
    state = extract_state_features(sim.grid, sim.dock)
    rows = []
    for a in legal:
        bi = a["block_idx"]
        wc = sim.count_clears_if_placed(bi, a["gx"], a["gy"])
        phi = extract_action_features(
            state,
            bi,
            a["gx"],
            a["gy"],
            sim.dock[bi]["shape"],
            wc,
            sim.grid.size,
            sim.grid,
            sim.dock,
            int(sim.dock[bi].get("color_idx", sim.dock[bi].get("colorIdx", 0))),
        )
        rows.append(phi)
    return state, np.stack(rows, axis=0) if rows else (state, np.zeros((0, PHI_DIM), dtype=np.float32))

File: web/src/bot/features.js
/**
 * 观测编码（state + action φ）：常数与归一化来自 shared/game_rules.json → FEATURE_ENCODING。
 * 改玩法规则一般只改 JSON；改特征结构时需改此处并与 Python observation 对齐，并重训模型。
 */
import { FEATURE_ENCODING } from '../gameRules.js';

const enc = FEATURE_ENCODING;
const AF = enc.almostFullLineRatio ?? 0.78;
const DOCK_SLOTS = enc.dockSlots ?? 3;
const AN = enc.actionNorm || {};
const MAX_GRID = enc.maxGridWidth ?? 8;
const DOCK_MASK_SIDE = enc.dockMaskSide ?? 5;
const STATE_SCALAR_DIM = enc.stateScalarDim ?? 15;

export const STATE_FEATURE_DIM = enc.stateDim;
export const ACTION_FEATURE_DIM = enc.actionDim;
export const PHI_DIM = enc.phiDim ?? enc.stateDim + enc.actionDim;

const _gridFlat = MAX_GRID * MAX_GRID;
const _dockFlat = DOCK_SLOTS * DOCK_MASK_SIDE * DOCK_MASK_SIDE;
const _expectState = STATE_SCALAR_DIM + _gridFlat + _dockFlat;
if (STATE_FEATURE_DIM !== _expectState) {
    throw new Error(
        `featureEncoding.stateDim=${STATE_FEATURE_DIM} 与标量+棋盘+待选区期望 ${_expectState} 不一致`
    );
}

/**
 * @param {import('../grid.js').Grid} grid
 */
function encodeGridOccupancy(grid) {
    const n = grid.size;
    const out = new Float32Array(_gridFlat);
    for (let y = 0; y < Math.min(n, MAX_GRID); y++) {
        for (let x = 0; x < Math.min(n, MAX_GRID); x++) {
            out[y * MAX_GRID + x] = grid.cells[y][x] !== null ? 1 : 0;
        }
    }
    return out;
}

/**
 * @param {number[][]} shape
 */
function encodeShapeMask(shape) {
    const h = shape.length;
    const w = h ? shape[0].length : 0;
    const side = DOCK_MASK_SIDE;
    const canvas = new Float32Array(side * side);
    if (!h || !w) {
        return canvas;
    }
    const offY = Math.max(0, Math.floor((side - h) / 2));
    const offX = Math.max(0, Math.floor((side - w) / 2));
    for (let py = 0; py < h; py++) {
        for (let px = 0; px < w; px++) {
            if (shape[py][px]) {
                const cy = offY + py;
                const cx = offX + px;
                if (cy >= 0 && cy < side && cx >= 0 && cx < side) {
                    canvas[cy * side + cx] = 1;
                }
            }
        }
    }
    return canvas;
}

/**
 * @param {{ shape: number[][], colorIdx: number, placed: boolean }[]} dock
 */
function encodeDockSpatial(dock) {
    const parts = [];
    for (let i = 0; i < DOCK_SLOTS; i++) {
        const b = dock[i];
        if (b && !b.placed) {
            parts.push(encodeShapeMask(b.shape));
        } else {
            parts.push(new Float32Array(DOCK_MASK_SIDE * DOCK_MASK_SIDE));
        }
    }
    const total = DOCK_SLOTS * DOCK_MASK_SIDE * DOCK_MASK_SIDE;
    const out = new Float32Array(total);
    let o = 0;
    for (const p of parts) {
        out.set(p, o);
        o += p.length;
    }
    return out;
}

/**
 * @param {import('../grid.js').Grid} grid
 * @param {{ shape: number[][], colorIdx: number, placed: boolean }[]} dock
 */
export function extractStateFeatures(grid, dock) {
    const n = grid.size;
    const area = n * n;
    let filled = 0;
    const rowFill = [];
    const colFill = [];

    for (let y = 0; y < n; y++) {
        let r = 0;
        for (let x = 0; x < n; x++) {
            if (grid.cells[y][x] !== null) {
                r++;
                filled++;
            }
        }
        rowFill.push(r / n);
    }

    for (let x = 0; x < n; x++) {
        let c = 0;
        for (let y = 0; y < n; y++) {
            if (grid.cells[y][x] !== null) {
                c++;
            }
        }
        colFill.push(c / n);
    }

    const maxRow = Math.max(...rowFill, 0);
    const minRow = rowFill.length ? Math.min(...rowFill) : 0;
    const maxCol = Math.max(...colFill, 0);
    const minCol = colFill.length ? Math.min(...colFill) : 0;

    let almostFullRows = 0;
    let almostFullCols = 0;
    for (const rf of rowFill) {
        if (rf >= AF && rf < 1) {
            almostFullRows++;
        }
    }
    for (const cf of colFill) {
        if (cf >= AF && cf < 1) {
            almostFullCols++;
        }
    }

    const unplaced = dock.filter((b) => !b.placed).length / DOCK_SLOTS;

    const scalars = new Float32Array([
        filled / area,
        maxRow,
        minRow,
        maxCol,
        minCol,
        almostFullRows / n,
        almostFullCols / n,
        unplaced,
        rowFill.reduce((a, b) => a + b, 0) / n,
        colFill.reduce((a, b) => a + b, 0) / n,
        stdDev(rowFill),
        stdDev(colFill),
        maxRow - minRow,
        maxCol - minCol,
        (almostFullRows + almostFullCols) / (2 * n)
    ]);
    if (scalars.length !== STATE_SCALAR_DIM) {
        throw new Error(`标量段长度 ${scalars.length} != stateScalarDim ${STATE_SCALAR_DIM}`);
    }

    const gridFlat = encodeGridOccupancy(grid);
    const dockFlat = encodeDockSpatial(dock);
    const out = new Float32Array(STATE_FEATURE_DIM);
    out.set(scalars, 0);
    out.set(gridFlat, STATE_SCALAR_DIM);
    out.set(dockFlat, STATE_SCALAR_DIM + _gridFlat);
    return out;
}

function stdDev(arr) {
    if (!arr.length) {
        return 0;
    }
    const m = arr.reduce((a, b) => a + b, 0) / arr.length;
    const v = arr.reduce((s, x) => s + (x - m) ** 2, 0) / arr.length;
    return Math.sqrt(v);
}

/**
 * @param {Float32Array} stateFeat
 * @param {number} blockIdx
 * @param {number} gx
 * @param {number} gy
 * @param {number[][]} shape
 * @param {number} wouldClear
 * @param {number} gridSize
 */
export function extractActionFeatures(stateFeat, blockIdx, gx, gy, shape, wouldClear, gridSize) {
    let cells = 0;
    for (let y = 0; y < shape.length; y++) {
        for (let x = 0; x < shape[y].length; x++) {
            if (shape[y][x]) {
                cells++;
            }
        }
    }
    const h = shape.length;
    const w = shape[0].length;
    const divB = AN.maxBlockIndex ?? 3;
    const divSh = AN.shapeSpan ?? 5;
    const divCells = AN.maxCells ?? 10;
    const divClr = AN.maxClearsHint ?? 5;
    const actionPart = new Float32Array([
        blockIdx / divB,
        gx / gridSize,
        gy / gridSize,
        w / divSh,
        h / divSh,
        cells / divCells,
        wouldClear / divClr
    ]);
    const out = new Float32Array(stateFeat.length + actionPart.length);
    out.set(stateFeat, 0);
    out.set(actionPart, stateFeat.length);
    return out;
}

/**
 * @param {import('./simulator.js').BlockBlastSimulator} sim
 * @returns {{ legal: { blockIdx: number, gx: number, gy: number }[], stateFeat: Float32Array, phiList: Float32Array[] }}
 */
export function buildDecisionBatch(sim) {
    const legal = sim.getLegalActions();
    const stateFeat = extractStateFeatures(sim.grid, sim.dock);
    const phiList = [];
    for (const a of legal) {
        const wouldClear = sim.countClearsIfPlaced(a.blockIdx, a.gx, a.gy);
        phiList.push(
            extractActionFeatures(
                stateFeat,
                a.blockIdx,
                a.gx,
                a.gy,
                sim.dock[a.blockIdx].shape,
                wouldClear,
                sim.grid.size
            )
        );
    }
    return { legal, stateFeat, phiList };
}
